File: src/zros_cpp/impl/service_servers_impl.cc
//
// Created by pairs on 7/16/18.
//

#include "service_servers_impl.h"
#include <sspdlog/sspdlog.h>
#include <zros/zros.h>
#include <zros/error.h>

namespace zros {

    //////////// IServersImpl ///////////////////
    bool IServersImpl::registerServer(const std::shared_ptr<IServiceServer> server) {
        std::lock_guard<std::mutex> lk (servers_mutex_);
        if (servers_.find(server->get_service_name()) != servers_.end()) {
            SSPD_LOG_WARNING << server->get_service_name() << " already register";
        }
        servers_[server->get_service_name()] = server;
        return true;
    }

    bool IServersImpl::unregisterServer(const std::string &service_name) {
        std::lock_guard<std::mutex> lk (servers_mutex_);
        servers_.erase(service_name);
        return true;
    }

    IServersImpl::IServersImpl(const std::string &address):service_address_(address) {

    }


    //////////// GrpcServersImpl ///////////////////
    const std::string& GrpcServersImpl::start() {
        if (!is_working_) {
            SSPD_LOG_INFO << "grpc server to work...";
            grpc::ServerBuilder builder;
            int select_port;
            builder.AddListeningPort(service_address_, grpc::InsecureServerCredentials(), &select_port);
            builder.RegisterService(this);
            grpc_server_ = std::move(builder.BuildAndStart());
            if (select_port == 0) {
                throw initialize_error("grpc server bind port failed");
            }
            if (service_address_ == "[::]:") {
                service_address_ = "localhost:" + std::to_string(select_port);
            }
            work_thread_ = std::make_shared<std::thread>([this]() {
                grpc_server_->Wait();
            });
            is_working_ = true;
        } else {
            SSPD_LOG_WARNING << "grpc server already work...";
        }
        return service_address_;
    }

    void GrpcServersImpl::stop() {
        grpc_server_->Shutdown();
        if (work_thread_) {
            work_thread_->join();
        }
    }

    GrpcServersImpl::GrpcServersImpl(const std::string &service_address):
            IServersImpl(service_address),
            is_working_(false){

    }

    grpc::Status GrpcServersImpl::InvokeService(grpc::ServerContext *context, const zros_rpc::ServiceRequest *request,
                                                zros_rpc::ServiceResponse *response) {
        SSPD_LOG_INFO << "receive rpc call " << request->service_name();
        auto existing_server = servers_.find(request->service_name());
        if (existing_server == servers_.end()) {
            response->mutable_status()->set_code(zros_rpc::Status_Code_FAILED_PRECONDITION);
            response->mutable_status()->set_details("There is no server for this service");
        } else {
            existing_server->second->invoke(request, response);
        }
        return grpc::Status::OK;
    }

    GrpcServersImpl::~GrpcServersImpl() {
        stop();
    }

}

File: src/zros_python/impl/service_servers_impl.py
# -*- coding: utf-8 -*-
""" impl service server management

data: 2018-07-31
author: pairs
"""

import abc
import logging
import threading
import zros_python.pb.zros_pb2_grpc as zpbg2
import zros_python.pb.zros_pb2 as zpb2
import grpc
from concurrent import futures

logger = logging.getLogger(__name__)


class IServersImpl(object):
    __metaclass__ = abc.ABCMeta

    def __init__(self, address):
        super(IServersImpl, self).__init__()
        self._address = address
        self._servers_mutex = threading.Lock()
        self._servers = {}

    def register_server(self, server):
        """

        :param server: ServiceServer
        :return:
        """
        service_name = server.get_service_name()
        with self._servers_mutex:
            if service_name in self._servers:
                logger.warning(u'register server: %s  have already register', service_name)
            else:
                self._servers[service_name] = server
                logger.info(u'register server: %s succeed', service_name)

    def unregister_server(self, service_name):
        """
        :param service_name:
        :return:
        """
        with self._servers_mutex:
            if service_name in self._servers:
                del self._servers[service_name]
                logger.info(u'unregister server: %s succeed', service_name)

    @abc.abstractmethod
    def start(self):
        pass

    @abc.abstractmethod
    def stop(self):
        pass


class GrpcServersImpl(IServersImpl, zpbg2.ServiceRPCServicer):
    """

    """
    def __init__(self, service_address, update_address_cb):
        """

        :param service_address:
        """
        super(GrpcServersImpl, self).__init__(service_address)
        self._grpc_server = None
        self._update_address_cb = update_address_cb

    def start(self):
        self._grpc_server = grpc.server(futures.ThreadPoolExecutor(max_workers=4))
        zpbg2.add_ServiceRPCServicer_to_server(self, self._grpc_server)
        if self._address == u'[::]:':
            port = self._grpc_server.add_insecure_port(u'[::]:')
            self._address = u'localhost:' + str(port)
        else:
            port = self._grpc_server.add_insecure_port(self._address)
        if port == 0:
            raise Exception(u'start service server failed')
        self._update_address_cb(self._address)
        self._grpc_server.start()
        logger.info(u'service servers run on %s', self._address)

    def stop(self):
        if self._grpc_server:
            self._grpc_server.stop(0)

    def __del__(self):
        self.stop()

    def InvokeService(self, request, context):
        """

        :param request:
        :param context:
        :return:
        """
        with self._servers_mutex:
            server = self._servers.get(request.service_name, None)
        if not server:
            response = zpb2.ServiceResponse()
            response.service_name = request.service_name
            response.status.code = zpb2.Status.UNKNOWN
            response.status.details = u"There is no server for the service"
            return response
        return server.invoke(request)
